sockeye/optimizers.py

```diff
@@ -0,0 +1,202 @@
+
+        # Now compute Eve's d term
+        d = 0.
+
+        # Eve rules from paper: compute f_hat and d based on last training batch objective
+        if self.use_batch_objective:
+            f_hat_prev, d_prev = prev_batch
+            f = self.batch_state.metric_val
+            if t > 1:
+                # The original paper has a typo in the algorithm here.  The following lines are re-
+                # written to reflect the actual logic presented in the authors' longer explanation.
+                if f <= f_hat_prev:
+                    delta_lo = 1. / (self.k_hi + 1.)
+                    delta_hi = 1. / (self.k_lo + 1.)
+                else:
+                    delta_lo = self.k_lo + 1.
+                    delta_hi = self.k_hi + 1.
+                # ^ End modified section ^
+                c = min(max(delta_lo, f / f_hat_prev), delta_hi)
+                f_hat = c * f_hat_prev
+                r = abs(f_hat - f_hat_prev) / min(f_hat, f_hat_prev)
+                d_batch = self.beta3 * d_prev + (1. - self.beta3) * r
+            else:
+                f_hat = f
+                d_batch = 1.
+            prev_batch[:] = [f_hat, d_batch]
+            d += d_batch
+
+        # Extension: compute f_hat and d based on last validation checkpoint objective
+        # Computation occurs once per checkpoint using the checkpoint number as t.  Prior to the
+        # first checkpoint, d = 1.
+        if self.use_checkpoint_objective:
+            checkpoint, f_hat_prev, d_prev = prev_checkpoint
+            # Only need to recompute if we've seen a new checkpoint since the previous batch update
+            if self.checkpoint_state and self.checkpoint_state.checkpoint != checkpoint:
+                checkpoint = self.checkpoint_state.checkpoint
+                f = self.checkpoint_state.metric_val
+                if checkpoint > 1:
+                    if f <= f_hat_prev:
+                        delta_lo = 1. / (self.k_hi + 1.)
+                        delta_hi = 1. / (self.k_lo + 1.)
+                    else:
+                        delta_lo = self.k_lo + 1.
+                        delta_hi = self.k_hi + 1.
+                    c = min(max(delta_lo, f / f_hat_prev), delta_hi)
+                    f_hat = c * f_hat_prev
+                    r = abs(f_hat - f_hat_prev) / min(f_hat, f_hat_prev)
+                    d_checkpoint = self.beta4 * d_prev + (1. - self.beta4) * r
+                else:
+                    f_hat = f
+                    d_checkpoint = 1.
+                prev_checkpoint[:] = [checkpoint, f_hat, d_checkpoint]
+            else:
+                d_checkpoint = d_prev
+            d += d_checkpoint
+
+        # Batch and checkpoint contribute equally when both are used
+        if self.use_batch_objective and self.use_checkpoint_objective:
+            d /= 2.
+
+        # Final weight update rule (Adam rule with extra d term)
+        weight[:] = weight - lr * mean / (d * sqrt(var) + self.epsilon)
```

Comment:
MXNet's standard Adam implements the update rule in C++, afaik (https://github.com/apache/incubator-mxnet/blob/master/python/mxnet/optimizer.py#L649). Could we reuse this? It seems that you are just scaling and shifting the variance term with d and eps here, so we could fold this into the var variable passed to adam_update and leave the update to the C++ backend.

Reply:
That would be ideal, though unfortunately it combines updating mean/var with the actual weight update.  We need to update var as it is, then scale by d for the weight update.